Makefile.pre.in

```diff
@@ -772,58 +772,112 @@ FROZEN_FILES_OUT = \
+#
+# This file is created by this rule, with an old timestamp.
+# It is deleted as soon as the state of the files changes.
+$(FROZEN_READY): .FORCE
+	@mkdir -p $(FROZEN_DATA)
+	@touch -t $(OLD_TIMESTAMP) $@
+	@rm -rf $(FROZEN_GENERATED)
+
+# This records that one or more frozen module .h files was built.  It also
+# prints an informative header when the first frozen module is handled.
+# It is run at the beginning of the recipe for each frozen module .h file.
+# Note that the rule for each module will print a progress marker on
+# the same line as the others.
+.PHONY: .generating-frozen-module
+.generating-frozen-module:
+	@if ( mkdir $(FROZEN_GENERATED) ) 2> /dev/null; then \
+		chmod 555 $(FROZEN_GENERATED); \
+		echo "=== Re-generating the frozen modules. ==="; \
+		echo "(For verbose output, run make regen-frozen.)"; \
+	else \
+		touch $(FROZEN_GENERATED); \
+	fi
+	@rm -f $(FROZEN_READY)
 
-Python/frozen_modules/posixpath.h: Programs/_freeze_module Lib/posixpath.py
-	Programs/_freeze_module posixpath $(srcdir)/Lib/posixpath.py $(srcdir)/Python/frozen_modules/posixpath.h
+# This should always run after all frozen module .h rules finish.
+# If at least one frozen module rules actually ran then this rule
+# finishes printing the line where the progress markers were printed.
+# This rule also removes all the other frozen-related stamp files.
+#
+# This file is created by the .generating-frozen-module rule if at least
+# one of the frozen module rules is run.  Otherwise it is created here,
+# with an old timestamp.  It is always an empty directory.
+# timestamp.  It is updated to "now" in each frozem module .h rule
+# if the module was actually updated.
+$(FROZEN_GENERATED): $(FROZEN_FILES_OUT) .FORCE
+	@if test -e $(FROZEN_GENERATED); then \
+		echo ""; \
+	else \
+		touch $@; \
+		chmod 555 $(FROZEN_GENERATED); \
+	fi
+	@touch -t $(OLD_TIMESTAMP) $@
+	@rm -f $(FROZEN_READY)
 
-Python/frozen_modules/os.h: Programs/_freeze_module Lib/os.py
-	Programs/_freeze_module os $(srcdir)/Lib/os.py $(srcdir)/Python/frozen_modules/os.h
+Programs/_freeze_module.o: Programs/_freeze_module.c Makefile
 
-Python/frozen_modules/site.h: Programs/_freeze_module Lib/site.py
-	Programs/_freeze_module site $(srcdir)/Lib/site.py $(srcdir)/Python/frozen_modules/site.h
+Programs/_freeze_module: Programs/_freeze_module.o $(LIBRARY_OBJS_OMIT_FROZEN) $(FROZEN_READY)
+	$(LINKCC) $(PY_CORE_LDFLAGS) -o $@ Programs/_freeze_module.o $(LIBRARY_OBJS_OMIT_FROZEN) $(LIBS) $(MODLIBS) $(SYSLIBS)
 
-Python/frozen_modules/stat.h: Programs/_freeze_module Lib/stat.py
-	Programs/_freeze_module stat $(srcdir)/Lib/stat.py $(srcdir)/Python/frozen_modules/stat.h
+have_frozen_files=$(wildcard $(FROZEN_FILES_OUT))
 
-Python/frozen_modules/__hello__.h: Programs/_freeze_module Lib/__hello__.py
-	Programs/_freeze_module __hello__ $(srcdir)/Lib/__hello__.py $(srcdir)/Python/frozen_modules/__hello__.h
+$(FROZEN_DATA)/%.h: Programs/_freeze_module
```

Comment:
FWIW, the commands in the block below this line give me the creeps (== I've never seen such make syntax before :-).

But perhaps we can define a [pattern rule](https://www.gnu.org/software/make/manual/html_node/Pattern-Examples.html#Pattern-Examples) that tells how to generate a frozen .h file from a Lib .py file? That won't change the output volume, but it will simplify the Makefile.